lib/constructBodyDigestSignature.js

```diff
@@ -0,0 +1,60 @@
+const crypto = require('crypto')
+
+const constructBodyDigestSignature = ({
+  headers: {
+    digest,
+    'content-type': contentType,
+    'content-encoding': contentEncoding
+  },
+  body
+}) => {
+  const { formattedAlgorithm, digestAlgorithm, digestValue } = parseBodyDigestHeader(digest)
+
+  const calculatedDigest = calculateDigest({
+    algorithm: formattedAlgorithm,
+    message: parseContentType(contentType, contentEncoding)(body),
+    encoding: lookupEncoding(formattedAlgorithm)
+  })
+
+  if (calculatedDigest !== digestValue) {
+    throw new Error('Calculated Digest does not match Digest from Request Header')
+  }
+
+  return `${digestAlgorithm}=${calculatedDigest}`
+}
+
+const parseBodyDigestHeader = (requestDigestHeader) => {
+  // TODO:(jeff-sexton) - Parse for multiple digests in comma separated list and select one
+  const [digestAlgorithm, digestValue] = requestDigestHeader.split(/=(.+)/)
+  const formattedAlgorithm = digestAlgorithm.toLowerCase().split('-').join('')
+
+  return { formattedAlgorithm, digestAlgorithm, digestValue }
+}
+
+const parseContentType = (requestContentTypeHeader, requestContentEncodingHeader) => {
+  // TODO:(jeff-sexton) - Add other media types and support for ContentEncoding
+  const mediaTypeFuncMap = {
+    'application/json': (resource) => JSON.stringify(resource),
+    default: () => { throw new Error('Unrecognized media-type') }
+  }
+
+  if (!requestContentTypeHeader) {
+    throw new Error('No Content-Type Header in request')
+  }
+
+  return mediaTypeFuncMap[requestContentTypeHeader.split(';')[0]] || mediaTypeFuncMap.default
+}
+
+const lookupEncoding = (algorithm) => {
+  const encodingAlgorithmMap = {
+    sha256: 'base64',
```

Comment:
These encodings can also be hex I believe. We should make this configurable in some way.

Reply:
My reading of this [IEFT draft](https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-digest-headers-02#section-5) led me to believe that sha256 and sha512 would always be base64. `UNIXsum`, `UNIXcksum`, `ID-SHA-512`, and `ID-SHA-256` might be different. 

The content encoding could also use hex but that would take place in `parseContentType`, not in the digest encoding of the algorithm output. 

I agree that we might want a way to specify the digest encoding, however. Otherwise, this lookup always needs to be up to date with the Internet Assigned Numbers Authority (IANA) registry. Where should this happen? Another default that can be overridden when the plugin is registered?

Reply:
Added `digestEncoding` property in the plugin options object. plugin default is still `base64` but this can now be overridden during plugin registration.

Reply:
👌